Inspect tab now uses an add/remove button

MorganMoon opened the pull request:
- choose which files to load in as asset bundles directly with add button
- remove button becomes active when user has an asset bundle selected, and when clicked, it removes it

My asset bundles don't use the same name as their .manifest file as needed before, and it's very possible others do the same. Also, I don't keep all my asset bundles in the same directory. Figured an add/remove system for this may solve these issues :)

This tool is great though - learned about it from Unite 👍

alffanclub commented:
I've looked it over and run it locally. I like the idea of handling more explicit bundle setups, but have some concerns. 
1. what if I have 100 bundles I want to look at?  this select-one-at-a-time technique is a bit cumbersome. Any ideas on how to get around that?
2. why are your bundles not named the same as the .manifest? Is it "append hash" or something else that's causing it? I don't entirely remember why I needed the .manifest check anyway. Could be we load all bundles in a given path and just exclude anything ending in .manifest.
3. some minor feedback on the code...
3.a. given how you're using m_Data.BundlePaths I'd make it a List<string> rather than an array. ArrayUtility.Add/Remove are both fairly expensive (Add calls Array.Resize under the covers, Remove actually creates a local List, removes the element, then converts that back to an array).  The array utility stuff is really best used when you have to be using an array and List isn't an option.
3.b. when doing disabled UI, it's better to use EditorGUI.DisabledScope() than GUI.enabled.  Search the browser's code base for DisabledScope to see how to use it.

Again, overall I like the idea and the implementation looks pretty good.  I just think that #1 above is a big concern to be worked out before integrating this. I'll ponder it and see if I can think of a way to make that scalable. Perhaps there's a way to support folders and explicit bundles.
Thanks for creating this and thanks for coming to Unite!
-Bill

MorganMoon commented:
Hey, thanks for the response! I added support for both folders & explicit bundles with a dialog popup. Couldn't find anything in unity that allowed mutli selection open file window (Maybe this could be pushed to the right people? https://feedback.unity3d.com/suggestions/editorutility-dot-openfilepanel-should-have-an-alternative-with-multiple-file-selection). I also fixed the array thing, and use EditorGUI.DisabledScope :) ... I have only really been using this tool for the Configure and Inspect tabs because I have already made my own asset bundle build setup, which renames them like so: {platform}-{assetBundle}.unity3d and not the manifest files. I also allowed multi selection so people could mass remove, and the inspect window has a little message that mutli inspection isn't supported. Thanks!

alffanclub commented on UnityEngine.AssetBundles/Editor/AssetBundleBrowser/InspectTab/AssetBundleInspectTab.cs:
```diff
@@ -133,39 +133,128 @@ private void OnGUIEditor()
+            if (m_Data.Contains(filePath))
+                return;
+
+            var bundleTestPath = this.LoadBundle(filePath);
```
this could be a non-trivial amount of time. Especially if the bundle is LZMA compressed as the entire thing will most likely have to be loaded. My vote is to always let the user add whatever they want (explicit bundle or folder of bundles), then upon inspection, either show the current data on the right side of the window, or a message saying "invalid bundle selected". That could also free us up to make the folder code a little more inclusive (not *only* look for files paired with a .manifest) since we'd be ok with false positives.

MorganMoon replied:
Makes sense, when we load a folder, just include everything? Would it be a good idea to have some extensions ignored? (.meta, .manifest, .dll, ect.)

alffanclub replied:
I would say yes to ignoring extensions in the "add folder" code, but not the "add single bundle". I've seen people with some very strange bundle naming preferences.  those three extensions you mention are probably good. maybe .cs as well, but it may be better to have too many than too few things show up.

alffanclub commented:
made one comment in-code for you to see. I'm going to be in and out of the office over the next two weeks, so my response time may slow down here a bit... I really like this idea/direction, but want to churn on it a little (I just may churn slowly).

alffanclub commented:
hey @MorganMoon  I've got some local changes I was making to the inspect tab, and I'd like to more easily iterate on this change of yours.  So I've created a new branch to merge into instead of master.  we can continue with changes here then merge this new branch to master once it's ready.  Thanks again for getting this started.
